### ISSUE TITLE
feat: use servodriver in WPT workflow

### ISSUE THREAD

**Issue by wusyong** [servo/servo#34683]
**Describe the new feature:**
This is a follow-up issue from #34550. It would be great if we could merge servo and servo driver WPT runners and run them in the WPT workflow. So tests that require testdriver.js can be tested instead of skipped. I'd like to help but I'm not too familiar with whole wptrunner setup yet.

And @jdm what does your [comment ](https://github.com/servo/servo/pull/34550#issuecomment-2531724812) about `--debugger` mean? Is it about opening GDB while running WPT?

**Comment by jdm** [comment]
Yes, the `--debugger` argument runs the browser binary under a debugger, and that feature is broken for the current servodriver implementation. #34551 is the other issue that makes servodriver unstable/slower than necessary right now.

**Comment by wusyong** [comment]
Another requirement is probably running in headless mode. I noticed it will open lots of window when test in batch.

**Comment by jdm** [comment]
I believe reftests in servodriver are also much slower than necessary in servodriver. I haven't yet done any investigation into why that's the case.

**Comment by jdm** [comment]
We also run into "Too many open files" errors under servodriver which causes tests to fail. I consistently hit this on `/_mozilla/mozilla/first-reflow-sheet-assert.html` when running `./mach test-wpt --product=servodriver tests/wpt/mozilla/tests/mozilla/ --processes=1`.

**Comment by wusyong** [comment]
> We also run into "Too many open files" errors under servodriver which causes tests to fail. I consistently hit this on `/_mozilla/mozilla/first-reflow-sheet-assert.html` when running `./mach test-wpt --product=servodriver tests/wpt/mozilla/tests/mozilla/ --processes=1`.

I noticed they didn't run in headless mode like default does. I assume opening too many real windows would cause too many files error. Not sure why servodriver didn't run them in headless mode though.

**Comment by jdm** [comment]
Headless or not does not make a difference. It's our proliferation of IPC channels and the persistent process that causes us to hit that limit.

**Comment by jdm** [comment]
https://github.com/servo/servo/issues/23905 has a lot of good prior art that we will probably need to resurrect to make this usable. Specifically the IPC router work that @gterzian spent a while investigating.

**Comment by mrobinson** [comment]
I am currently working on removing the leak of IPC channels that are used for fetch cancellers. This should help the situation, but probably not fix it entirely.

**Comment by jdm** [comment]
Yes, the other big one that affects webdriver in particular is https://github.com/servo/servo/blob/21975784de6b43e439dd9a5539fa42bffbba12c8/components/script/dom/window.rs#L2099, because the presence of a webdriver server means that screenshots are a possibility.

**Comment by jdm** [comment]
Interestingly, there was no change in behaviour after #34833. We still panic with "Too many open files" on the same test. The backtrace is:
```
0:32.49 pid:74031 Can't read file: Os { code: 24, kind: Uncategorized, message: "Too many open files" } (thread Constellation, at ports/servoshell/resources.rs:80)
 0:34.87 pid:74031    0: backtrace::backtrace::libunwind::trace
 0:34.87 pid:74031              at /Users/jdm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.74/src/backtrace/libunwind.rs:116:5
 0:34.87 pid:74031       backtrace::backtrace::trace_unsynchronized
 0:34.87 pid:74031              at /Users/jdm/.cargo/registry/src/index.crates.io-6f17d22bba15001f/backtrace-0.3.74/src/backtrace/mod.rs:66:5
 0:34.87 pid:74031    1: <servoshell::backtrace::Print as core::fmt::Debug>::fmt
 0:34.87 pid:74031              at /Users/jdm/src/alt-servo/ports/servoshell/backtrace.rs:65:13
 0:34.88 pid:74031    2: core::fmt::rt::Argument::fmt
 0:34.88 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/fmt/rt.rs:177:76
 0:34.88 pid:74031       core::fmt::write
 0:34.88 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/fmt/mod.rs:1186:21
 0:34.90 pid:74031    3: std::io::Write::write_fmt
 0:34.90 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/io/mod.rs:1839:15
 0:34.90 pid:74031    4: servoshell::backtrace::print
 0:34.90 pid:74031              at /Users/jdm/src/alt-servo/ports/servoshell/backtrace.rs:18:5
 0:34.90 pid:74031    5: servoshell::panic_hook::panic_hook
 0:34.90 pid:74031              at /Users/jdm/src/alt-servo/ports/servoshell/panic_hook.rs:40:17
 0:34.92 pid:74031    6: core::ops::function::Fn::call
 0:34.92 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/ops/function.rs:79:5
 0:34.97 pid:74031    7: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/alloc/src/boxed.rs:2468:9
 0:34.97 pid:74031       std::panicking::rust_panic_with_hook
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:809:13
 0:34.97 pid:74031    8: std::panicking::begin_panic_handler::{{closure}}
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:674:13
 0:34.97 pid:74031    9: std::sys::backtrace::__rust_end_short_backtrace
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/sys/backtrace.rs:170:18
 0:34.97 pid:74031   10: rust_begin_unwind
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:665:5
 0:34.97 pid:74031   11: core::panicking::panic_fmt
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/panicking.rs:74:14
 0:34.97 pid:74031   12: core::result::unwrap_failed
 0:34.97 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/result.rs:1700:5
 0:34.98 pid:74031   13: core::result::Result<T,E>::expect
 0:34.98 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/result.rs:1061:23
 0:34.98 pid:74031   14: <servoshell::resources::ResourceReader as embedder_traits::resources::ResourceReaderMethods>::read
 0:34.98 pid:74031              at /Users/jdm/src/alt-servo/ports/servoshell/resources.rs:80:9
 0:34.99 pid:74031   15: embedder_traits::resources::read_bytes
 0:34.99 pid:74031              at /Users/jdm/src/alt-servo/components/shared/embedder/resources.rs:30:5
 0:35.05 pid:74031   16: <net::image_cache::ImageCacheImpl as net_traits::image_cache::ImageCache>::new
 0:35.05 pid:74031              at /Users/jdm/src/alt-servo/components/net/image_cache.rs:421:26
 0:35.07 pid:74031   17: constellation::pipeline::UnprivilegedPipelineContent::start_all
 0:35.07 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/pipeline.rs:510:36
 0:35.07 pid:74031   18: constellation::pipeline::Pipeline::spawn
 0:35.07 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/pipeline.rs:310:21
 0:35.12 pid:74031   19: constellation::constellation::Constellation<STF,SWF>::new_pipeline
 0:35.12 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:969:22
 0:35.12 pid:74031   20: constellation::constellation::Constellation<STF,SWF>::handle_new_top_level_browsing_context
 0:35.12 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:3014:9
 0:35.13 pid:74031   21: constellation::constellation::Constellation<STF,SWF>::handle_webdriver_msg
 0:35.13 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:4533:17
 0:35.13 pid:74031   22: constellation::constellation::Constellation<STF,SWF>::handle_request_from_compositor
 0:35.13 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:1432:17
 0:35.13 pid:74031   23: constellation::constellation::Constellation<STF,SWF>::handle_request
 0:35.13 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:1202:45
 0:35.13 pid:74031   24: constellation::constellation::Constellation<STF,SWF>::run
 0:35.13 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:777:13
 0:35.13 pid:74031   25: constellation::constellation::Constellation<STF,SWF>::start::{{closure}}
 0:35.13 pid:74031              at /Users/jdm/src/alt-servo/components/constellation/constellation.rs:764:17
 0:35.14 pid:74031   26: std::sys::backtrace::__rust_begin_short_backtrace
 0:35.14 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/sys/backtrace.rs:154:18
 0:35.16 pid:74031   27: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
 0:35.16 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/thread/mod.rs:538:17
 0:35.19 pid:74031   28: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
 0:35.19 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/panic/unwind_safe.rs:272:9
 0:35.20 pid:74031   29: std::panicking::try::do_call
 0:35.20 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:557:40
 0:35.20 pid:74031   30: ___rust_try
 0:35.20 pid:74031   31: std::panicking::try
 0:35.20 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panicking.rs:520:19
 0:35.20 pid:74031       std::panic::catch_unwind
 0:35.20 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/panic.rs:358:14
 0:35.20 pid:74031       std::thread::Builder::spawn_unchecked_::{{closure}}
 0:35.20 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/thread/mod.rs:537:30
 0:35.24 pid:74031   32: core::ops::function::FnOnce::call_once{{vtable.shim}}
 0:35.24 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/core/src/ops/function.rs:250:5
 0:35.24 pid:74031   33: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:35.24 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/alloc/src/boxed.rs:2454:9
 0:35.24 pid:74031       <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
 0:35.24 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/alloc/src/boxed.rs:2454:9
 0:35.24 pid:74031       std::sys::pal::unix::thread::Thread::new::thread_start
 0:35.24 pid:74031              at /rustc/90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf/library/std/src/sys/pal/unix/thread.rs:105:17
 0:35.24 pid:74031   34: __pthread_joiner_wake
```
which is very interesting. We're creating one image cache per script thread, which reads the rippy.png resource from the embedder. Per #33672 that means two consumed file descriptors per script thread that is created during the tests, and that apparently happens _a lot_.

**Comment by jdm** [comment]
Oof—every time webdriver navigates to about:blank, it is allocated a new script thread.

**Comment by mrobinson** [comment]
Resources should probably be cached per-process once read, but this worries me a bit. I Looking at the pattern used by the ImageCache, `generate_image_key` seems to be the only function creating an IPC channel and it is immediately used and dropped. I wonder if we are somehow leaking FDs after dropping the channel -- or maybe it's just that we are consuming them too fast for the OS.

**Comment by jdm** [comment]
I've been experimenting with [try runs](https://github.com/servo/servo/actions/runs/12729842540/job/35482128335) using servodriver. The first problem I note is that our webxr tests fails because we only register a mock device during startup (https://github.com/servo/servo/blob/2115b6a6db33de8c86b6c4899926c7f0923825a0/ports/servoshell/desktop/embedder.rs#L51-L52) but the webdriver test harness only sets the pref after startup is already complete.

**Comment by jdm** [comment]
#35667 is a new blocker to turning on servodriver by default.

**Comment by jdm** [comment]
https://github.com/servo/servo/actions/runs/17508457294 is a WPT run with `--product servodriver --headless` that turned up a combination of recent regressions and exciting new failures. As well as fixing #39175 and #39180, which are responsible for vast numbers of failing tests, we will need to automate updating the expected results for any test that uses testdriver. Once all that work is done, it will be easier to see what additional unique failures remain.
